dd-java-agent/instrumentation/netty-4.0/src/main/java/datadog/trace/instrumentation/netty40/AttributeKeys.java

```
@@ -1,27 +1,41 @@
 package datadog.trace.instrumentation.netty40;
 
-import datadog.trace.bootstrap.WeakMap;
+import datadog.trace.bootstrap.WeakCache;
 import datadog.trace.bootstrap.instrumentation.api.AgentSpan;
 import datadog.trace.context.TraceScope;
 import datadog.trace.instrumentation.netty40.client.HttpClientTracingHandler;
 import datadog.trace.instrumentation.netty40.server.HttpServerTracingHandler;
 import io.netty.util.AttributeKey;
-import java.util.Map;
+import java.util.Iterator;
+import java.util.ServiceLoader;
+import java.util.concurrent.Callable;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 
 public class AttributeKeys {
+  private static WeakCache<ClassLoader, ConcurrentMap<String, AttributeKey<?>>> newWeakCache() {
+    final Iterator<WeakCache.Provider> providers =
+        ServiceLoader.load(WeakCache.Provider.class, null).iterator();
+    if (providers.hasNext()) {
+      final WeakCache.Provider provider = providers.next();
+      if (providers.hasNext()) {
+        throw new IllegalStateException(
+            "Only one implementation of WeakCache.Provider suppose to be in classpath");
+      }
+      return provider.newWeakCache();
+    }
+    throw new IllegalStateException("Can't load implementation of WeakCache.Provider");
+  }
 
-  private static final WeakMap<ClassLoader, Map<String, AttributeKey<?>>> map =
-      WeakMap.Implementation.DEFAULT.get();
-
-  private static final WeakMap.ValueSupplier<ClassLoader, Map<String, AttributeKey<?>>>
-      mapSupplier =
-          new WeakMap.ValueSupplier<ClassLoader, Map<String, AttributeKey<?>>>() {
-            @Override
-            public Map<String, AttributeKey<?>> get(final ClassLoader ignored) {
-              return new ConcurrentHashMap<>();
-            }
-          };
+  private static final WeakCache<ClassLoader, ConcurrentMap<String, AttributeKey<?>>> map =
```

I don't think this class can use a cache semantic.  The reason this exists is because netty will throw an error if an attribute of the same name is created more than once.  (See the comment in the class.)

---

Maybe just cache without time eviction ? only by weak reference expiration

---

Conceptually, I don't think of WeakReference expiration as eviction per se.  From the standpoint of retrieving something from the Map/Cache, WeakReference expiration is an observable difference.  Only the side effect on memory is observable.

---

> I don't think this class can use a cache semantic. The reason this exists is because netty will throw an error if an attribute of the same name is created more than once. (See the comment in the class.)

That doesn't necessarily mean that cache won't work, but it depends on the nature of the error.  If Netty fails fast and in recoverable way, then we could simply handle the exception rather than introducing a side structure.